src/codegen/codegen_cpp_visitor.hpp

```diff
@@ -1677,8 +1677,8 @@ class CodegenCppVisitor: public visitor::ConstAstVisitor {
      * \param target_printer A printer defined outside this visitor to be used for the code
      *                       generation
      */
-    CodegenCppVisitor(std::string mod_filename,
-                      std::string float_type,
+    CodegenCppVisitor(const std::string& mod_filename,
+                      const std::string& float_type,
```

Comment:
The more modern pattern is:
```
CodegenCppVisitor(std::string mod_filename, ...)
  : mod_filename(std::move(mod_filename))
```

Reply:
I guess this only applies for the strings/objects that get assigned straight to some member variable?
For example in this case:
```
CodegenCppVisitor(std::string mod_filename,
                      const std::string& output_dir,
                      std::string float_type,
                      const bool optimize_ionvar_copies,
                      const std::string& extension,
                      const std::string& wrapper_ext)
        : target_printer(new CodePrinter(output_dir + "/" + mod_filename + extension))
        , wrapper_printer(new CodePrinter(output_dir + "/" + mod_filename + wrapper_ext))
        , printer(target_printer)
        , mod_filename(std::move(mod_filename))
        , float_type(std::move(float_type))
        , optimize_ionvar_copies(optimize_ionvar_copies) {}
```
it doesn't make sense to do the same with `output_dir` since we only use it in `new CodePrinter(output_dir...`?

Reply:
Also `output_dir` is used twice, so moving would be a bug. Unrelated, the `new` should be replaced with `std::make_shared<CodePrinter>(output_dir + "/" + ...)`.

Finally and just as unrelated: the pattern `var + "/" + "bla"` is very error prone. If `output_dir` is empty, which is usually synonymous with `"."` we end up with `/bla` instead of `./bla`.

Reply:
As discussed in Slack, `std::make_shared` is better than `new CodePrinter` because that way we make the allocation for the `CodePrinter` and its reference counter at the same point.
Since in some cases we share the printer it's not possible to use a `unique_ptr` for the `*printer` variables.
Regarding the paths, I see what you mean but since this would include a lot of changes (probably the best way to handle this is using `std::filesystem`) I would do this in a separate PR. I will create an issue about that